dowhy/causal_identifiers/efficient_backdoor.py

```diff
@@ -0,0 +1,375 @@
+import networkx as nx
+import numpy as np
+
+
+EXCEPTION_NO_ADJ = "An adjustment set formed by observable variables does not exist"
+EXCEPTION_COND_NO_OPT = "Conditions to guarantee the existence of an optimal adjustment set are not satisfied"
+
+
+class EfficientBackdoor:
+    """
+    Implements methods for finding optimal (efficient) backdoor sets.
+    """
+
+    def __init__(self, graph, conditional_node_names=None, costs=None):
+        """
+        :param graph: CausalGraph
+            A causal graph.
+        :param costs: list
+            A list with non-negative costs associated with variables in the graph. Only used
+            for estimatand_type='non-parametric-ate' and method_name='efficient-mincost-adjustment'. If
+            not costs are provided by the user, and method_name='efficient-mincost-adjustment', costs
+            are assumed to be equal to one for all variables in the graph. The structure of the list should
+            be of the form [(node, {"cost": x}) for node in nodes].
+        :param conditional_node_names: list
```

Comment:
ah, I finally understood this parameter. If I get it correctly, it is exactly the `effect_modifiers` parameter that can passed to the CausalModel class when initializing or to the estimate_effect method. Currently, this parameter is used to decide which variables to condition on to compute the CATE. 

So in that sense, conditional_nodes are used in the same way as effect_modifiers that determine what the user is going to estimate CATE on after identifying the effect using efficient backdoor. how do you think about renaming it to `effect_modifiers` to be consistent with the [CausalEstimator](https://github.com/py-why/dowhy/blob/main/dowhy/causal_estimator.py) class (see init method)?

Reply:
These conditional nodes indicate variables that are used to decide when to apply treatment. For instance, in the hypothetical situation described in the notebook, a researcher might be interested in estimating the effect that a given policy to allocate treatment (warm up exercises), based on a given set of variables (previous injury and team motivation). Such a policy could for example be, apply treatment (warm up =1) only when the patient has previous injury = 1 and team motivation <= 6. But it could be any other function of these two variables; it could also be a randomised function. The backdoor sets we compute allow you to estimate the mean of the outcome in this counterfactual world, regardless of how the policy actually works (as long as certain positivity conditions hold). 

In contrast, effect modifiers, at least as I understand them, are variables that define strata of the population that one believes somehow interact with the effect of the treatment, and it is of interest to understand how this effect changes across different strata. You could also be interested in seeing how the effect of a policy (as described above) on an outcome varies across subsets of the population; I'm not sure how the optimality theory we have developed would work though.

Thus, I believe these two sets of variables play very different roles.

Reply:
I explained the role of the conditional nodes in the example more thoroughly [here](https://github.com/py-why/dowhy/pull/549/commits/35ccf7b236b06a8e97731e9107fb886aab8adf65#diff-a091d9c0ebb6f97cbf4a55d010dc43f3095300529377911336fa6d7866eb2228R79)

Reply:
got it, I am starting to understand this better.

While not necessary for this PR, it will still be good to understand how these `conditional_nodes` may be useful later when you have collected the data and want to estimate the treatment's effect? Is it correct to think of them as the parents of the treatment node (that could also be parents of outcome Y)? And would the policy function become the analogue of the propensity score model if it is a randomized function that outputs probabilities?

The reason I'm asking is that we've been thinking about what the the right abstractions to ask people to enter information about the graph. Currently dowhy asks `common_causes`, but there is a proposal to include `outcome_only_causes` and `treatment_only_causes` separately to specify the nodes that are parents of one but not the other. Looks like onditional_nodes denotes all parents of treatment here, so that would be `treatment_only_causes union common_causes`? Am I understanding this correctly?

Reply:
Hi Amit! Here go some extra clarifications.

Consider the example in the notebook, and suppose the treatment, whether someone does warm up exercises, is binary, and so is the outcome, whether they have injuries after playing sports. 

Suppose a researcher is interested in the causal effect that prescribing players to do warm-up exercises has on injury. If the hypothetical intervention is to make all players (regardless of their individual characteristics) do warm up exercises, then the causal effect can be quantified via the ATE, the difference in the expected outcomes in two “parallel” worlds, one in which everyone does warm up exercises (regardless of the player’s individual characteristics), the other in which no one does warm up exercises. In this case the conditional nodes are the empty set (there are none), because the intervention is carried out with no regard to individual characteristics.

Now, perhaps a different, individualised, intervention is of interest. That is, the researcher might be interested in estimating the injury probability in a counterfactual world, in which players are prescribed to perform warm up exercises according to some of their characteristics. One such policy could for example be, apply treatment (warm up =1) only when the patient has previous injury = 1 and team motivation <= 6. Another one could be, if the person has tissue disorder then flip a fair coin and if it lands heads have them to warm up exercises, if it lands tails then have them don’t do warm up exercises. Notice that this latter treatment rule is randomised. For the first case, the conditional nodes are previous injury and team motivation, for the second case the only conditional node is tissue disorder.

Thus, conditional nodes can be any set of non-descendants of treatment (including the empty set); they need not be parents of the treatment.  Changing the conditional nodes changes what the target estimand is.

Suppose now we built a graph, computed an adjustment set (possibly for an individualised treatment rule), and have collected data on the treatment, the conditional nodes and the outcome. Then we can estimate the expectation of the counterfactual outcome under our treatment rule of interest using any of the usual estimation strategies (IPW, outcome regression, AIPW, etc). To make matters more concrete, let Y be the outcome, A the treatment,  L=tissue disorder, and Z be the adjustment set we got from running any of the algorithms on the graph. Let pi(A |L) be the randomised treatment rule based on tissue disorder we described before, that is pi(A=1 |L)= 0.5 if L=tissue_disorder=1 and pi(A=1 |L)= 0 if  L=tissue_disorder=0. Let f^hat(A| Z) the estimated propensity score. Then an estimate of the expected outcome in the counterfactual world in which people are allocated to treatment according to the pi(A |L) treatment rule is given by 
average ( Y * pi(A|L) /  f^hat(A| Z) ). This is the IPW estimator; the other ones are defined analogously. Notice that if pi(A|L) did not depend on L, say pi(A|L)=1 when A=1, regardless of the value of L, then this is just the usual formula for the IPW.

Reply:
got it, thank you so much for the detailed reply (and for being patient with my repeated queries).